pkgdb/src/lock-flake-installable.cc

```diff
@@ -0,0 +1,320 @@
+    },
+    lockFlags );
+  debugLog(
+    nix::fmt( "locked installable: '%s'", installable->what().c_str() ) );
+
+
+  auto lockedUrl = installable->getLockedFlake()->flake.lockedRef.to_string();
+  debugLog( nix::fmt( "locked url: '%s'", lockedUrl ) );
+
+  auto flakeDescription = installable->getLockedFlake()->flake.description;
+
+  auto cursor = installable->getCursor( *state );
+
+  auto lockedAttrPath = cursor->getAttrPathStr();
+  debugLog( nix::fmt( "locked attr path: '%s'", lockedAttrPath ) );
+
+  debugLog( nix::fmt( "locked outputs: '%s'",
+                      installable->extendedOutputsSpec.to_string() ) );
+
+  // check if the output is a derivation (not a just a store path)
+  if ( ! cursor->isDerivation() )
+    {
+      auto v = cursor->forceValue();
+      throw nix::Error(
+        "expected flake output attribute '%s' to be a derivation but found %s",
+        lockedAttrPath,
+        nix::showType( v ) );
+    }
+
+  // read the drv path
+  std::string derivation;
+  {
+    auto derivationCursor
+      = cursor->findAlongAttrPath( nix::parseAttrPath( *state, "drvPath" ) );
+    if ( ! derivationCursor ) { return EXIT_FAILURE; }
+    derivation = ( *derivationCursor )->getStringWithContext().first;
+  }
+
+  // map output names to their store paths
+  std::map<std::string, std::string> outputs;
+  std::vector<std::string>           outputNames;
+  {
+    auto maybe_outputs_cursor
+      = cursor->findAlongAttrPath( nix::parseAttrPath( *state, "outputs" ) );
+    if ( ! maybe_outputs_cursor ) { return EXIT_FAILURE; }
+    outputNames = ( *maybe_outputs_cursor )->getListOfStrings();
+
+    for ( auto output : outputNames )
+      {
+        auto outputCursor = cursor->findAlongAttrPath(
+          nix::parseAttrPath( *state, output + ".outPath" ) );
+        if ( ! outputCursor ) { return EXIT_FAILURE; }
+        auto outputValue = ( *outputCursor )->getStringWithContext();
+        outputs[output]  = outputValue.first;
+      }
+  }
+
+  // determine outputs to install in the following order:
+  // 1. extendedOutputsSpec (`<installable>^out,man`, `<installable>^*`))
+  // 2. `meta.outputsToInstall` field
+  // 3. first output in the `outputs` field
+  std::set<std::string> outputsToInstall;
+  {
+    auto outputSpec = std::visit(
+      overloaded {
+        [&]( const nix::ExtendedOutputsSpec::Default & ) -> nix::OutputsSpec
+        {
+          std::set<std::string> outputsToInstall;
+          auto metaOutputsToInstallCursor = cursor->findAlongAttrPath(
+            nix::parseAttrPath( *state, "meta.outputsToInstall" ) );
+          if ( metaOutputsToInstallCursor )
+            {
+              for ( auto output :
+                    ( *metaOutputsToInstallCursor )->getListOfStrings() )
+                {
+                  outputsToInstall.insert( output );
+                }
+            }
+          else if ( ! outputNames.empty() )
+            {
+              outputsToInstall.insert( outputNames.front() );
```

Comment:
We currently have no way to specify outputs for catalog packages. Those currently behave like `^*`. Conversely we can't specify "only the default output" for flake installables.
Thus behaviour between the two will differ unless we adjust the default for either.

I believe this is currently non-blocking as `outputs-to-install` is not being used by the builder: https://github.com/flox/flox/issues/1363

Reply:
Thanks for highlighting. FTR the Nix-expression-based buildenv implementation I was mentioning the other day honors outputs-to-install, so that might be another reason to consider trying it out at this time (in addition to the dual-buildenv renderings of runtime and developer symlink trees).

Reply:
I agree it's nonblocking, but I almost think `outputs_spec` should be a separate field in the lockfile. We've had some back and forth on whether we should default to `outputsToInstall` or `^*`. If we store both `outputs_spec` and `outputsToInstall` in a single field, we can't change the default behavior without re-scraping. `^out` would be indistinguishable from `meta.outputsToInstall = [ "out" ]`, but when it comes to defaulting, we probably want to always respect the former, but not necessarily the latter

Reply:
ok i added `requested-outputs-to-install` and `output-names` to allow us do any sort of defaulting downstream instead